From cc4ca8ab5316aa36224871b2418aab77c7fc86f9 Mon Sep 17 00:00:00 2001
From: someone5678 <nemui3353@gmail.com>
Date: Wed, 31 May 2023 22:33:20 +0900
Subject: [PATCH 3/3] base: Hide ADB and developer setting enable status [1/2]

* Some apps check whether ADB or developer setting is enabled
* Return 0 to selected apps when those apps getInt
  ADB_ENABLED, ADB_WIFI_ENABLED, DEVELOPMENT_SETTINGS_ENABLED
* Code is adapted and stripped down from cutout force full screen

Ref:
https://github.com/LineageOS/android_lineage-sdk/blob/lineage-16.0/sdk/src/java/org/lineageos/internal/applications/LongScreen.java
https://github.com/yaap/frameworks_base/commit/232260f192effc9fb21a54eb444601179b24a26d

Signed-off-by: someone5678 <nemui3353@gmail.com>

base: HideDevStatus: move to getIntForUser to cover more scenarios

* Adapt to current project as we want to set status individually per-app
  for case apps require ADB or Dev settings for their functionality

e.g. SystemUI Tuner by Zachary Wander
https://play.google.com/store/apps/details?id=com.zacharee1.systemuituner&hl=en_US&gl=US

Ref:
https://github.com/RisingTechOSS/android_frameworks_base/commit/77be7be42060dc5119c6dccef874c9a439aed210

Change-Id: Ibbf92c7b1f1181d2508cb22ff2fc2017f9e59706
Co-authored-by: minaripenguin <minaripenguin@users.noreply.github.com>
Signed-off-by: someone5678 <nemui3353@gmail.com>

base: HideDeveloperStatus: Query all apps installed by any user [1/2]

* And set value for all user

Change-Id: Ibd2a5dfab832e6bdc6d074ce41401f4905602152

base: HideDeveloperStatusUtils: Avoid NPE with mApps

* https://paste.evolution-x.org/P9Lj8g.

* Avoid NPE

* Clean-up and handle exceptions

Signed-off-by: someone5678 <someone5678@users.noreply.github.com>

core: Improve hide adb/developer spoofing

basically does what this module does:
https://github.com/accelforce/DevOptsHide/blob/master/app/src/main/java/net/accelf/devoptshide/HideDevOpts.kt

to reduce calls and unnecessary looping for adb/developer hidden apps, intercept every method and do an early return instead of looping over get methods also add getString since some apps like Gcash is weird enough to use getString for an integer value.

https://github.com/accelforce/DevOptsHide/blob/master/app/src/main/java/net/accelf/devoptshide/HideDevOpts.kt#L63

also fix a recursion

test
- without patch: open gcash with developer and usb debugging enabled. gcash detects and crashes
- with patch: same procedure, gcash doesn't detect anything, gcash works as intended

Change-Id: I8214c2f2642d38e97ab8d193fde8b5dcddc58a57
Signed-off-by: rmp22 <195054967+rmp22@users.noreply.github.com>
Signed-off-by: therealmharc <therealmharc@gmail.com>
---
 core/java/android/provider/Settings.java      |  63 ++++++++++
 .../util/custom/HideDeveloperStatusUtils.java | 118 ++++++++++++++++++
 .../validators/SecureSettingsValidators.java  |   1 +
 3 files changed, 182 insertions(+)
 create mode 100644 core/java/com/android/internal/util/custom/HideDeveloperStatusUtils.java

diff --git a/core/java/android/provider/Settings.java b/core/java/android/provider/Settings.java
index b5a4d706ea75..29eaf6388ac8 100644
--- a/core/java/android/provider/Settings.java
+++ b/core/java/android/provider/Settings.java
@@ -111,6 +111,8 @@ import android.widget.Editor;
 import com.android.internal.annotations.GuardedBy;
 import com.android.internal.util.Preconditions;
 
+import com.android.internal.util.custom.HideDeveloperStatusUtils;
+
 import java.io.IOException;
 import java.lang.annotation.ElementType;
 import java.lang.annotation.Retention;
@@ -3836,6 +3838,9 @@ public final class Settings {
         @UnsupportedAppUsage
         public String getStringForUser(ContentResolver cr, String name,
                 final @CanBeCURRENT @UserIdInt int userId) {
+            if (HideDeveloperStatusUtils.shouldHideDevStatus(cr, cr.getPackageName(), name)) {
+                return "0" /* Disabled */;
+            }
             final boolean isSelf = (userId == UserHandle.myUserId());
             final AttributionSource attributionSource = cr.getAttributionSource();
             final int deviceId = android.permission.flags.Flags.deviceAwarePermissionApisEnabled()
@@ -4578,6 +4583,9 @@ public final class Settings {
          * @return the corresponding value, or null if not present
          */
         public static String getString(ContentResolver resolver, String name) {
+            if (HideDeveloperStatusUtils.shouldHideDevStatus(resolver, resolver.getPackageName(), name)) {
+                return "0" /* Disabled */;
+            }
             return getStringForUser(resolver, name, resolver.getUserId());
         }
 
@@ -4585,6 +4593,9 @@ public final class Settings {
         @UnsupportedAppUsage
         public static String getStringForUser(ContentResolver resolver, String name,
                 @CanBeCURRENT @UserIdInt int userId) {
+            if (HideDeveloperStatusUtils.shouldHideDevStatus(resolver, resolver.getPackageName(), name)) {
+                return "0" /* Disabled */;
+            }
             if (MOVED_TO_SECURE.contains(name)) {
                 Log.w(TAG, "Setting " + name + " has moved from android.provider.Settings.System"
                         + " to android.provider.Settings.Secure, returning read-only value.");
@@ -4782,6 +4793,9 @@ public final class Settings {
          * or not a valid integer.
          */
         public static int getInt(ContentResolver cr, String name, int def) {
+            if (HideDeveloperStatusUtils.shouldHideDevStatus(cr, cr.getPackageName(), name)) {
+                return 0 /* Disabled */;
+            }
             return getIntForUser(cr, name, def, cr.getUserId());
         }
 
@@ -4789,6 +4803,9 @@ public final class Settings {
         @UnsupportedAppUsage
         public static int getIntForUser(ContentResolver cr, String name, int def,
                 @CanBeCURRENT @UserIdInt int userId) {
+            if (HideDeveloperStatusUtils.shouldHideDevStatus(cr, cr.getPackageName(), name)) {
+                return 0 /* Disabled */;
+            }
             String v = getStringForUser(cr, name, userId);
             return parseIntSettingWithDefault(v, def);
         }
@@ -4813,6 +4830,9 @@ public final class Settings {
          */
         public static int getInt(ContentResolver cr, String name)
                 throws SettingNotFoundException {
+            if (HideDeveloperStatusUtils.shouldHideDevStatus(cr, cr.getPackageName(), name)) {
+                return 0 /* Disabled */;
+            }
             return getIntForUser(cr, name, cr.getUserId());
         }
 
@@ -4821,6 +4841,9 @@ public final class Settings {
         public static int getIntForUser(ContentResolver cr, String name,
                 @CanBeCURRENT @UserIdInt int userId)
                 throws SettingNotFoundException {
+            if (HideDeveloperStatusUtils.shouldHideDevStatus(cr, cr.getPackageName(), name)) {
+                return 0 /* Disabled */;
+            }
             String v = getStringForUser(cr, name, userId);
             return parseIntSetting(v, name);
         }
@@ -7845,6 +7868,9 @@ public final class Settings {
          * @return the corresponding value, or null if not present
          */
         public static String getString(ContentResolver resolver, String name) {
+            if (HideDeveloperStatusUtils.shouldHideDevStatus(resolver, resolver.getPackageName(), name)) {
+                return "0" /* Disabled */;
+            }
             return getStringForUser(resolver, name, resolver.getUserId());
         }
 
@@ -7852,6 +7878,9 @@ public final class Settings {
         @UnsupportedAppUsage
         public static String getStringForUser(ContentResolver resolver, String name,
                 @CanBeCURRENT @UserIdInt int userId) {
+            if (HideDeveloperStatusUtils.shouldHideDevStatus(resolver, resolver.getPackageName(), name)) {
+                return "0" /* Disabled */;
+            }
             if (MOVED_TO_GLOBAL.contains(name)) {
                 Log.w(TAG, "Setting " + name + " has moved from android.provider.Settings.Secure"
                         + " to android.provider.Settings.Global.");
@@ -8080,6 +8109,9 @@ public final class Settings {
          * or not a valid integer.
          */
         public static int getInt(ContentResolver cr, String name, int def) {
+            if (HideDeveloperStatusUtils.shouldHideDevStatus(cr, cr.getPackageName(), name)) {
+                return 0 /* Disabled */;
+            }
             return getIntForUser(cr, name, def, cr.getUserId());
         }
 
@@ -8087,6 +8119,9 @@ public final class Settings {
         @UnsupportedAppUsage
         public static int getIntForUser(ContentResolver cr, String name, int def,
                 @CanBeCURRENT @UserIdInt int userId) {
+            if (HideDeveloperStatusUtils.shouldHideDevStatus(cr, cr.getPackageName(), name)) {
+                return 0 /* Disabled */;
+            }
             String v = getStringForUser(cr, name, userId);
             return parseIntSettingWithDefault(v, def);
         }
@@ -8111,6 +8146,9 @@ public final class Settings {
          */
         public static int getInt(ContentResolver cr, String name)
                 throws SettingNotFoundException {
+            if (HideDeveloperStatusUtils.shouldHideDevStatus(cr, cr.getPackageName(), name)) {
+                return 0 /* Disabled */;
+            }
             return getIntForUser(cr, name, cr.getUserId());
         }
 
@@ -8118,6 +8156,9 @@ public final class Settings {
         public static int getIntForUser(ContentResolver cr, String name,
                 @CanBeCURRENT @UserIdInt int userId)
                 throws SettingNotFoundException {
+            if (HideDeveloperStatusUtils.shouldHideDevStatus(cr, cr.getPackageName(), name)) {
+                return 0 /* Disabled */;
+            }
             String v = getStringForUser(cr, name, userId);
             return parseIntSetting(v, name);
         }
@@ -14010,6 +14051,13 @@ public final class Settings {
         @Readable
         public static final String HIDE_APPLIST = "hide_applist";
 
+        /**
+         * Control whether to hide ADB and Developer settings enable status.
+         * @hide
+         */
+        @Readable
+        public static final String HIDE_DEVELOPER_STATUS = "hide_developer_status";
+
         /**
          * Keys we no longer back up under the current schema, but want to continue to
          * process when restoring historical backup datasets.
@@ -20072,6 +20120,9 @@ public final class Settings {
          * @return the corresponding value, or null if not present
          */
         public static String getString(ContentResolver resolver, String name) {
+            if (HideDeveloperStatusUtils.shouldHideDevStatus(resolver, resolver.getPackageName(), name)) {
+                return "0" /* Disabled */;
+            }
             return getStringForUser(resolver, name, resolver.getUserId());
         }
 
@@ -20079,6 +20130,9 @@ public final class Settings {
         @UnsupportedAppUsage(maxTargetSdk = Build.VERSION_CODES.R, trackingBug = 170729553)
         public static String getStringForUser(ContentResolver resolver, String name,
                 int userHandle) {
+            if (HideDeveloperStatusUtils.shouldHideDevStatus(resolver, resolver.getPackageName(), name)) {
+                return "0" /* Disabled */;
+            }
             if (MOVED_TO_SECURE.contains(name)) {
                 Log.w(TAG, "Setting " + name + " has moved from android.provider.Settings.Global"
                         + " to android.provider.Settings.Secure, returning read-only value.");
@@ -20294,6 +20348,9 @@ public final class Settings {
          * or not a valid integer.
          */
         public static int getInt(ContentResolver cr, String name, int def) {
+            if (HideDeveloperStatusUtils.shouldHideDevStatus(cr, cr.getPackageName(), name)) {
+                return 0 /* Disabled */;
+            }
             String v = getString(cr, name);
             return parseIntSettingWithDefault(v, def);
         }
@@ -20318,6 +20375,9 @@ public final class Settings {
          */
         public static int getInt(ContentResolver cr, String name)
                 throws SettingNotFoundException {
+            if (HideDeveloperStatusUtils.shouldHideDevStatus(cr, cr.getPackageName(), name)) {
+                return 0 /* Disabled */;
+            }
             String v = getString(cr, name);
             return parseIntSetting(v, name);
         }
@@ -22683,6 +22743,9 @@ public final class Settings {
         @RequiresPermission(Manifest.permission.READ_DEVICE_CONFIG)
         public static String getString(@NonNull String name) {
             ContentResolver resolver = getContentResolver();
+            if (HideDeveloperStatusUtils.shouldHideDevStatus(resolver, resolver.getPackageName(), name)) {
+                return "0" /* Disabled */;
+            }
             return sNameValueCache.getStringForUser(resolver, name, resolver.getUserId());
         }
 
diff --git a/core/java/com/android/internal/util/custom/HideDeveloperStatusUtils.java b/core/java/com/android/internal/util/custom/HideDeveloperStatusUtils.java
new file mode 100644
index 000000000000..534f05f7dc9f
--- /dev/null
+++ b/core/java/com/android/internal/util/custom/HideDeveloperStatusUtils.java
@@ -0,0 +1,118 @@
+package com.android.internal.util.custom;
+
+import android.content.ContentResolver;
+import android.content.Context;
+import android.os.SystemProperties;
+import android.os.UserHandle;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+import android.provider.Settings;
+
+public class HideDeveloperStatusUtils {
+    private static final Set<String> settingsToHide =
+        new HashSet<>(
+            Arrays.asList(
+                Settings.Global.ADB_ENABLED,
+                Settings.Global.ADB_WIFI_ENABLED,
+                Settings.Global.DEVELOPMENT_SETTINGS_ENABLED
+            ));
+
+    enum Action {
+        ADD,
+        REMOVE,
+        SET
+    }
+
+    private static boolean isBootCompleted() {
+        return SystemProperties.getBoolean("sys.boot_completed", false);
+    }
+
+    public static boolean shouldHideDevStatus(ContentResolver cr, String packageName, String name) {
+        if (cr == null || packageName == null || name == null || !isBootCompleted()) {
+            return false;
+        }
+        
+        if (!settingsToHide.contains(name)) return false;
+
+        Set<String> apps = getApps(cr);
+        if (apps.isEmpty()) {
+            return false;
+        }
+
+        return apps.contains(packageName);
+    }
+
+    private static Set<String> getApps(Context context) {
+        if (context == null) {
+            return new HashSet<>();
+        }
+
+        return getApps(context.getContentResolver());
+    }
+
+    private static Set<String> getApps(ContentResolver cr) {
+        if (cr == null) {
+            return new HashSet<>();
+        }
+
+        String apps = Settings.Secure.getString(cr, Settings.Secure.HIDE_DEVELOPER_STATUS);
+        if (apps != null && !apps.isEmpty() && !apps.equals(",")) {
+            return new HashSet<>(Arrays.asList(apps.split(",")));
+        }
+
+        return new HashSet<>();
+    }
+
+    private static void putAppsForUser(
+            Context context, String packageName, int userId, Action action) {
+        if (context == null || userId < 0) {
+            return;
+        }
+
+        final Set<String> apps = getApps(context);
+        switch (action) {
+            case ADD:
+                apps.add(packageName);
+                break;
+            case REMOVE:
+                apps.remove(packageName);
+                break;
+            case SET:
+                // Don't change
+                break;
+        }
+
+        Settings.Secure.putStringForUser(
+                context.getContentResolver(),
+                Settings.Secure.HIDE_DEVELOPER_STATUS,
+                String.join(",", apps),
+                userId);
+    }
+
+    public void addApp(Context mContext, String packageName, int userId) {
+        if (mContext == null || packageName == null || userId < 0) {
+            return;
+        }
+
+        putAppsForUser(mContext, packageName, userId, Action.ADD);
+    }
+
+    public void removeApp(Context mContext, String packageName, int userId) {
+        if (mContext == null || packageName == null || userId < 0) {
+            return;
+        }
+
+        putAppsForUser(mContext, packageName, userId, Action.REMOVE);
+    }
+
+    public void setApps(Context mContext, int userId) {
+        if (mContext == null || userId < 0) {
+            return;
+        }
+
+        putAppsForUser(mContext, null, userId, Action.SET);
+    }
+}
diff --git a/packages/SettingsProvider/src/android/provider/settings/validators/SecureSettingsValidators.java b/packages/SettingsProvider/src/android/provider/settings/validators/SecureSettingsValidators.java
index 1d7e4248df8e..83ba04114610 100644
--- a/packages/SettingsProvider/src/android/provider/settings/validators/SecureSettingsValidators.java
+++ b/packages/SettingsProvider/src/android/provider/settings/validators/SecureSettingsValidators.java
@@ -564,5 +564,6 @@ public class SecureSettingsValidators {
         VALIDATORS.put(Secure.NETWORK_TRAFFIC_AUTOHIDE, BOOLEAN_VALIDATOR);
         VALIDATORS.put(Secure.NETWORK_TRAFFIC_UNITS, new InclusiveIntegerRangeValidator(0, 4));
         VALIDATORS.put(Secure.HIDE_APPLIST, ANY_STRING_VALIDATOR);
+        VALIDATORS.put(Secure.HIDE_DEVELOPER_STATUS, ANY_STRING_VALIDATOR);
     }
 }
-- 
2.43.0

